staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -414,13 +401,22 @@ func (qs *queueSet) chooseQueueIndexLocked(hashValue uint64, descr1, descr2 inte
 	bestQueueIdx := -1
 	bestQueueLen := int(math.MaxInt32)
 	// the dealer uses the current desired number of queues, which is no larger than the number in `qs.queues`.
-	qs.dealer.Deal(hashValue, func(queueIdx int) {
-		thisLen := len(qs.queues[queueIdx].requests)
-		klog.V(7).Infof("QS(%s): For request %#+v %#+v considering queue %d of length %d", qs.config.Name, descr1, descr2, queueIdx, thisLen)
-		if thisLen < bestQueueLen {
-			bestQueueIdx, bestQueueLen = queueIdx, thisLen
-		}
-	})
+	if qs.config.Dealer != nil {
+		qs.config.Dealer.Deal(hashValue, func(queueIdx int) {
+			if queueIdx < 0 || queueIdx >= len(qs.queues) {
+				return
+			}
+			thisLen := len(qs.queues[queueIdx].requests)
+			klog.V(7).Infof("QS(%s): For request %#+v %#+v considering queue %d of length %d", qs.config.Name, descr1, descr2, queueIdx, thisLen)
+			if thisLen < bestQueueLen {
+				bestQueueIdx, bestQueueLen = queueIdx, thisLen
+			}
+		})
+	}
+	if bestQueueIdx < 0 {
+		panic(fmt.Sprintf("dealer dealt no valid cards; config=%#+v", qs.config))
```

Comment:
I asked for this and I think it is the right thing to do: explicit panic is better than silently dropping fairness. An even better thing to do would be to make the change I asked for in another comment so that it is impossible to construct a `qs` with a bad dealer, then this code can go back like it was.

Panics should get caught and re-panic'd on the main goroutine, which will cause them to be logged. I asked for that behavior in yet another comment.